### ISSUE TITLE
aarch32 stdenv should hide the architecture of the builder machine

### ISSUE THREAD

**Issue by volth** [NixOS/nixpkgs#60103]
## Issue description

```aarch64``` build box is able to compile for ```aarch32```
https://discourse.nixos.org/t/new-aarch64-and-armv7l-linux-omg-builders/1010

The problem is: build scripts of many software (openssl, libffi, perl, python, ...) are not aware of ```aarch64``` machines running ```aarch32``` code, detect then the host system is ```aarch64``` and then do insane things like compiling ```aarch64``` assemler files into ```aarch32``` build.

It was not fixed on time expecting @Ericson2314 proposal for "always cross compile" (https://github.com/NixOS/nixpkgs/issues/21471) is to be accepted soon, but it did not happen.

So the idea is to fix ```aarch32``` ```stdenv``` so it would hide from 3rd-party scripts the true architecture of the builder machine  (by providing fake ```uname``` and tricks like that)

**Comment by matthewbauer** [comment]
Faking uname just hides the problem. These build scripts should have a way to override autodetection. Usually this is available as either a configure or make flag.

**Comment by volth** [comment]
Sometimes it is not easy: for example, ```python``` has vendored ```libffi``` and the flag is to be passed to ./configure of ```libffi```.
So it is not just one more flags in ```configureFlags```, it is a patch difficult to maintain, because Hydra and OfBorg do not build for ```aarch32```

**Comment by volth** [comment]
Removal of ```uname``` from ```stdenv``` (at least when sandbox is on) might be a step forward.

Then all the derivation which relied on ```uname``` will have to be explicitly parameterized with the architecture and that will be maintained

**Comment by Ericson2314** [comment]
Yeah my prefixed compilers PR wasn't ready, but we could just pass `--build` and `--host` always (for the derivations that support it). That seems good to me. I forget if there were other issues.

CC @lheckemann

**Comment by samueldr** [comment]
I'm a bit worried that this cannot be tested well, as even in x86_64 land we're having issues with CPU feature detection done wrong, and it goes mostly unnoticed. What can we do to better ensure that the outputs will not have armv8 specific instructions, and/or aarch64 outputs (when and if armv8 is not always strictly equivalent to aarch64*)?

<sup>
*I am still a bit unclear as to whether armv8 in 32 bit mode is strictly armv7 equivalent or not, could not find documentation corroborating or discrediting that.
</sup>

**Comment by volth** [comment]
It is interesting to run aarch32 qemu on that server, to check if it can use hardware virtualisation.
In case of success, the virtual machine can be used to produce armv7 builds without leaking armv8 features of the builder.

**Comment by samueldr** [comment]
I can assert that yes, QEMU with KVM can be used on that machine type. Yes, I was able to produce useful armv7 builds with that setup, including a UEFI armv7 iso image from a WIP branch. In case that knowledge is useful.

**Comment by lheckemann** [comment]
Something something `setarch linux32`. It won't be sufficient to get differentiated `armv[5678]` builds, but should at least get around configure scripts thinking they should build for aarch64. Question would be where to integrate it.

@samueldr that sounds good! Even more useful than the knowledge that you managed to do that would be the knowledge how you managed to do that ;)

**Comment by samueldr** [comment]
I think we verified in the past that `setarch linux32` isn't enough; I don't remember how, but it works differently than how 32 bit personality works for x86_64. Though I would be glad to be found to be wrong.

**Comment by Ericson2314** [comment]
@volth's own #59855 does somewhat contradict this goal!

**Comment by volth** [comment]
For `aarch32` builder on more powerful `aarch64` hardware it might be worth to try to run `aarch64` kernel with `aarch32` userland (it works on x86 https://github.com/NixOS/nixpkgs/issues/60103)

**Comment by lheckemann** [comment]
Alternatively: https://github.com/grahamc/packet-nix-builder/pull/2

>run aarch64 kernel with aarch32 userland

Last I checked, build systems will still autodetect aarch64 based on uname, causing things to explode again.

**Comment by samueldr** [comment]
If memory serves, the reason it works with i686 vs. x86_64 is because there is special casing in the kernel for `setarch` and x86 families. The equivalent doesn't exist for ARM families. <sup>\[Citation needed\]</sup>

This is why it's not as simple as `pkgsAArch32.hello`.

**Comment by lheckemann** [comment]
There _is_ some sort of equivalent — at least `personality(PER_LINUX32)` (can be used on the command line with the `linux32` command) causes uname to report `armv8l` instead of `aarch64`. Nix doesn't do this though, and I'm not sure it would be sufficient to make things build armv7-compatible code. Probably worth trying though!

**Comment by cleverca22** [comment]
for x86-64 specifically, `-cpu Haswell` gives you more SSE features, while still not auto-detecting all features the host can provide

it also fixes the fact that `-cpu host` hard-breaks qemu if `/dev/kvm` doesnt exist

**Comment by samueldr** [comment]
Does that actually work with kvm when CPU features are discovered? I seem to recall trying to test the `sage` issue with qemu and could not reproduce the issue as long as it was running on top of kvm. I'm not sure if I was able to reproduce using emulation, I don't remember if I had the patience to wait for it.

**Comment by Ericson2314** [comment]
CC @Gaelan
